pkg/vm/engine/tae/containers/cn_vector_adapter.go

```diff
@@ -0,0 +1,640 @@
+// Copyright 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containers
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/RoaringBitmap/roaring"
+	"github.com/RoaringBitmap/roaring/roaring64"
+	"github.com/matrixorigin/matrixone/pkg/common/mpool"
+	"github.com/matrixorigin/matrixone/pkg/compress"
+	cnNulls "github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	cnVector "github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/vm/engine/tae/common"
+	"io"
+	"unsafe"
+)
+
+type CnTaeVector[T any] struct {
+	downstreamVector *cnVector.Vector
+
+	// Nullable is mainly used while Marshalling & Unmarshalling
+	isNullable bool
+
+	// Mpool is mostly defined within DN vector. So reusing the same approach for simplicity.
+	mpool *mpool.MPool
+
+	// isAllocatedFromMpool is used with Allocated() & ResetWithData()
+	// When ResetWithData(bytes) is called, we are not allocating using the vector's Mpool.
+	// isAllocatedFromMpool  is used to track that. So when we call Allocated() after ResetWithData(), we should get Zero.
+	isAllocatedFromMpool bool
+}
+
+func NewCnTaeVector[T any](typ types.Type, nullable bool, opts ...Options) *CnTaeVector[T] {
+	vec := CnTaeVector[T]{
+		downstreamVector: cnVector.New(typ),
+		isNullable:       nullable,
+	}
+
+	// nullable
+	if nullable {
+		vec.downstreamVector.Nsp = cnNulls.NewWithSize(0)
+	}
+
+	// setting mpool variables
+	var alloc *mpool.MPool
+	if len(opts) > 0 {
+		alloc = opts[0].Allocator
+	}
+	if alloc == nil {
+		alloc = common.DefaultAllocator
+	}
+	vec.mpool = alloc
+	vec.isAllocatedFromMpool = false
+
+	return &vec
+}
+
+func (vec *CnTaeVector[T]) Get(i int) any {
```

Comment:
Do you want to replace `any` with `T`? 

Something similar to the current CN implementation.
```go
func GetValueAt[T types.FixedSizeT](v *Vector, idx int64) T {
	return MustTCols[T](v)[idx]
}
```

I think it should be possible. `Get(i)` has around 277 usages, and in every place, we are doing explicit typecasting. I will refactor the vector to use Get[T](i) so that the calls can be directly forwarded to the CN Vector.

![image](https://user-images.githubusercontent.com/9638314/217728991-62d2a6c0-bdf9-4d76-836d-fb4060b3c89d.png)

Please note: I will include these mass refactoring changes in a separate PR to keep the current PR concise for reability.

Reply:
As @XuPeng-SH suggested yesterday, I don't think we can replace `any` with `T`.
Reason: T can't be of type `types.FixedSizeT` as it doesn't include `types.Null{}`